File: leetcode/785.py
"""
785. Is Graph Bipartite?
https://leetcode.com/problems/is-graph-bipartite/
dfs
"""
class Solution:
    def isBipartite(self, graph: List[List[int]]) -> bool:
        n = len(graph)
        colors = [-1] * n
        
        for i in range(n):
            if colors[i] == -1:
                if not self.dfs(graph, colors, i, 0):
                    return False
        return True
    
    def dfs(self, graph, colors, node, color):
        n = len(graph)
        colors[node] = color
        for e in graph[node]:
            if colors[e] != -1 and colors[e] == color:
                return False
            elif colors[e] == -1 and not self.dfs(graph, colors, e, 1 - color):
                return False
        return True

File: leetcode/785.cpp
/**
 * 785. Is Graph Bipartite?
 * https://leetcode.com/problems/is-graph-bipartite/
 * dfs
 */
class Solution {
public:
    /**
     * @param graph: the given undirected graph
     * @return:  return true if and only if it is bipartite
     */
    bool isBipartite(vector<vector<int>> &graph) {
        // Write your code here
        int n = graph.size();
        vector<int> colors(n, -1);
        for (int i = 0; i < n; ++i) {
            if (colors[i] == -1)
                if (!dfs(i, graph, colors, 0))
                    return false;
        }
        return true;
    }
    
    bool dfs(int index, vector<vector<int>> &graph, vector<int> &colors, int color) {
        colors[index] = color;
        for (auto &e: graph[index]) {
            if (colors[e] != -1 && colors[e] == color) return false;
            if (colors[e] == -1 && !dfs(e, graph, colors, 1 - color)) return false;
        }
        return true;
    }
};